### PR TITLE
fancysets changes

### LINKED ISSUE #7067: factor_list() error Python3
```
Following function call generate an error in Python3 but work well in Python2.

>>> factor_list(x*(x+y))

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/thilinarmtb/GSoC/sympy/sympy/polys/polytools.py", line 5675, in factor_list
    return _generic_factor_list(f, gens, args, method='factor')
  File "/home/thilinarmtb/GSoC/sympy/sympy/polys/polytools.py", line 5423, in _generic_factor_list
    fp = _sorted_factors(fp, method)
  File "/home/thilinarmtb/GSoC/sympy/sympy/polys/polytools.py", line 5322, in _sorted_factors
    return sorted(factors, key=key)
TypeError: unorderable types: list() < int()


Similarly following calls generate the same error:
[1] factor_list(y*(x+y)),  factor_list(x*y + y**2)
[2] factor_list(x*(x+y+z)), factor_list(x**2+x*y+x*z)
[3] factor_list((x + 1)*(x + y))

But the following don't:
[4] factor_list(x*(x**2 + y**2)), factor_list(x**3 + x*y**2) 
[5] factor_list(x**3 - x),  factor_list(x*(x**2 - 1))
[6] factor_list(x**2 + x*y + x + y)
[7] factor_list((x + z)*(x + y)), factor_list(x**2 + x*z + y*x + y*z)

Here [6] is mathematically same as [3]. But only [3] gives the error.
```
Original issue for #7067: http://code.google.com/p/sympy/issues/detail?id=3968
Original author: https://code.google.com/u/115610317529675726783/
Original owner: https://code.google.com/u/101069955704897915480/

### LINKED ISSUE #10956: ImageSet intersection not working correctly on specific input.
I am getting this issue when working on an issue.
```
In [1]: s = imageset(Lambda(x,x*pi),S.Integers)

In [2]: t = imageset(Lambda(x,x*pi+pi/2),S.Integers)

In [17]: s
Out[17]: {pi*x | x in Integers()}

In [18]: t
Out[18]:
        pi
{pi*x + -- | x in Integers()}
        2

In [19]: s.intersect(t)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-19-2462679590bc> in <module>()
----> 1 s.intersect(t)

c:\Users\Nitin\Desktop\Sympy\sympy\sympy\sets\sets.pyc in intersect(self, other)

    106
    107         """
--> 108         return Intersection(self, other)
    109
    110     def intersection(self, other):

c:\Users\Nitin\Desktop\Sympy\sympy\sympy\sets\sets.pyc in __new__(cls, *args, **
kwargs)
   1390         # Reduce sets using known rules
   1391         if evaluate:
-> 1392             return Intersection.reduce(args)
   1393
   1394         return Basic.__new__(cls, *args)

c:\Users\Nitin\Desktop\Sympy\sympy\sympy\sets\sets.pyc in reduce(args)
   1561                 new_args = False
   1562                 for t in args - set((s,)):
-> 1563                     new_set = s._intersect(t)
   1564                     # This returns None if s does not know how to inters
ect
   1565                     # with t. Returns the newly intersected set otherwis
e

c:\Users\Nitin\Desktop\Sympy\sympy\sympy\sets\fancysets.pyc in _intersect(self,
other)
    326                 a, b = Dummy('a'), Dummy('b')
    327                 f, g = f.subs(n, a), g.subs(m, b)
--> 328                 solns_set = diophantine(f - g)
    329                 if solns_set == set():
    330                     return EmptySet()

c:\Users\Nitin\Desktop\Sympy\sympy\sympy\solvers\diophantine.pyc in diophantine(
eq, param)
     90
     91         if eq_type in ["linear", "homogeneous_ternary_quadratic", "gener
al_pythagorean"]:
---> 92             if merge_solution(var, var_t, solution) != ():
     93                 sols.add(merge_solution(var, var_t, solution))
     94

c:\Users\Nitin\Desktop\Sympy\sympy\sympy\solvers\diophantine.pyc in merge_soluti
on(var, var_t, solution)
    115     l = []
    116
--> 117     if None in solution:
    118         return ()
    119

TypeError: argument of type 'NoneType' is not iterable

In [4]: t.intersect(s)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-fad60302fdd4> in <module>()
----> 1 t.intersect(s)

.......................................................................
c:\Users\Nitin\Desktop\Sympy\sympy\sympy\solvers\diophantine.pyc in merge_soluti
on(var, var_t, solution)
    115     l = []
    116
--> 117     if None in solution:
    118         return ()
    119

TypeError: argument of type 'NoneType' is not iterable

In [5]: s-t
Out[5]:
                                   pi
{pi*x | x in Integers()} \ {pi*x + -- | x in Integers()}
                                   2

In [6]: t-s
Out[6]:
        pi
{pi*x + -- | x in Integers()} \ {pi*x | x in Integers()}
        2
```
diop_base is returning None which is the problem. But when i do complement then it works.

### PR DESCRIPTION
**Pull request by smichr**
closes #7067
closes #10956

### CODE REVIEW

File: sympy/solvers/diophantine.py [hunk 1]
```diff
@@ -55,12 +55,12 @@ def diophantine(eq, param=symbols("t", integer=True)):
     >>> from sympy.solvers.diophantine import diophantine
     >>> from sympy.abc import x, y, z
     >>> diophantine(x**2 - y**2)
-    set([(-t_0, -t_0), (t_0, -t_0)])
+    set([(t_0, -t_0), (t_0, t_0)])
```
Comment (smichr) [inline sympy/solvers/diophantine.py]: after this change the next run gave

```
3941Expected:

3942    set([(t_0, -t_0), (t_0, t_0)])

3943Got:

3944    set([(-t_0, -t_0), (t_0, -t_0)])

3945**********************************************************************

3946File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/sympy-1.0.1.dev0-py3.3.egg/sympy/solvers/diophantine.py", line 60, in sympy.solvers.diophantine.diophantine

3947Failed example:

3948    diophantine(x*(2*x + 3*y - z))

3949Expected:

3950    set([(0, n1, n2), (3*t - z, -2*t + z, z)])

3951Got:

3952    set([(0, n1, n2), (t_0, -t_1, 2*t_0 - 3*t_1)])

3953**********************************************************************

3954File "/home/travis/virtualenv/python3.3.5/lib/python3.3/site-packages/sympy-1.0.1.dev0-py3.3.egg/sympy/solvers/diophantine.py", line 62, in sympy.solvers.diophantine.diophantine

3955Failed example:

3956    diophantine(x**2 + 3*x*y + 4*x)

3957Expected:

3958    set([(0, n1), (3*t - 4, -t)])

3959Got:

3960    set([(0, n1), (3*t_0 - 4, -t_0)])
```

so there is some non-determinism present. This sign problem can be fixed by making the first element of a tuple canonically signed and then changing the other appropriately. I wonder if `clear_coefficients` could be used to reduce the first element to a symbol to give all the constants on the second (and latter) elements....

Comment (smichr) [inline sympy/solvers/diophantine.py]: check on how the elements are sorted, too, because line 3950 looks like 1st and 3rd are reversed

Comment (smichr) [inline sympy/solvers/diophantine.py]: both variants are solutions:

```
>>> eq=(x*(2*x + 3*y - z))
>>> var('t_0 t_1 t_2')
(t_0 ,t_1 ,t_2 ) 
>>> A=(3*t - z, -2*t + z, z)
>>> B=(t_0, -t_1, 2*t_0 - 3*t_1)
>>> eq.subs(list(zip((x,y,z),A)))
0 
>>> eq.subs(list(zip((x,y,z),B)))
0
```

File: sympy/solvers/diophantine.py [hunk 2]
```diff
@@ -334,8 +336,8 @@ def classify_diop(eq):
         x, y = var[:2]
         diop_type = "cubic_thue"
 
-        for term in [x**3, x**2*y, x*y**2, y**3, Integer(1)]:
-            if term not in coeff.keys():
+        for term in [x**3, x**2*y, x*y**2, y**3, S.One]:
+            if term not in coeff:
                 coeff[term] == Integer(0)
```
Comment (smichr) [inline sympy/solvers/diophantine.py]: change Integer(0) to S.Zero

File: sympy/solvers/diophantine.py [hunk 3]
```diff
@@ -614,13 +607,14 @@ def base_solution_linear(c, a, b, t=None):
...
     d = igcd(a, igcd(b, c))
+    assert d > 0
     a = a // d
     b = b // d
     c = c // d
 
     if c == 0:
         if t != None:
-            return (b*t , -a*t)
+            return (b*t , -a*t) if b > 0 else (-b*t, a*t)
```
Comment (smichr) [inline sympy/solvers/diophantine.py]: may not be necessary with changes of how `factor_terms` results are used.

File: sympy/solvers/diophantine.py [hunk 4]
```diff
@@ -55,12 +55,12 @@ def diophantine(eq, param=symbols("t", integer=True)):
     >>> from sympy.solvers.diophantine import diophantine
     >>> from sympy.abc import x, y, z
     >>> diophantine(x**2 - y**2)
-    set([(-t_0, -t_0), (t_0, -t_0)])
+    set([(t_0, -t_0 + 1), (t_0, t_0 + 1)])
```
Comment (smichr) [inline sympy/solvers/diophantine.py]: how is that `+1` getting there if we are only ignoring sign? This is clearly wrong. Is it part of the `merge_terms` (or whatever it is named) process?

**Comment by smichr** [comment]
@jksuom @asmeurer , I think I got the problem fixed. I'm still a bilt puzzled by what I was seeing since that was an old change that I don't think is represented anywhere in the current PR. Perhaps the force update and online editing of this PR caused a glitch.

File: sympy/solvers/diophantine.py [hunk 5]
```diff
@@ -532,35 +484,27 @@ def _diop_linear(var, coeff, param):
...
-                k = 1
-                param_index = params.index(args[j]) + 1
-            else: #args[j] is an Integer
+        for j, arg in enumerate(Add.make_args(c)):
+            k, p = arg.as_coeff_Mul()
+            if p.is_Integer:   # arg is an Integer
                 # example: 5 -> k = 5
-                k = args[j]
                 param_index = 0
+                if not p:
```
Comment (smichr) [inline sympy/solvers/diophantine.py]: for this reason, `as_coeff_Mul` should probably always return (0, 0) for 0. Check codebase to see how it's used.

File: sympy/solvers/diophantine.py [hunk 6]
```diff
@@ -678,7 +623,7 @@ def divisible(a, b):
     """
     Returns `True` if ``a`` is divisible by ``b`` and `False` otherwise.
     """
-    return igcd(int(a), int(b)) == abs(int(b))
+    return not a % b
```
Comment (smichr) [inline sympy/solvers/diophantine.py]: the Diophantine routines should be careful to always work in integers, and to make only the public function return SymPy types. Right now, however, too many functions are public, I think.

File: sympy/sets/fancysets.py [hunk 7]
```diff
@@ -153,21 +155,38 @@ def _boundary(self):
...
 
+        a = Wild('a', exclude=[n])
+        b = Wild('b', exclude=[n])
         match = expr.match(a*n + b)
-        if match[a] is S.One and match[b].is_integer:
-            expr = expr - match[b]
+        if match:
+            # make sign on x positive
+            if match[a].is_negative:
+                expr = -expr
```
Comment (jksuom) [inline sympy/sets/fancysets.py]: This appears to modify the image. We are expected to have the constant  b  in the image for  n = 0. After this change we seem to get  -b  instead. (Unless  b  and  -b  differ by a multiple of  a.)

**Comment by smichr** [comment]
@jksuom , you were right. The behavior you were seeing was consistent with the behavior in master but was wrong. While f(x) and f(-x) will cover the same integers, f(x) and -f(x) will not. I now only make two simplifications (which are also a bit more robust).

File: sympy/polys/polytools.py [hunk 8]
```diff
@@ -5622,12 +5622,12 @@ def _sorted_factors(factors, method):
         def key(obj):
             poly, exp = obj
             rep = poly.rep.rep
-            return (exp, len(rep), rep)
+            return (exp, len(rep), len(poly.gens), rep)
```
Comment (smichr) [inline sympy/polys/polytools.py]: this needs a test that fails in master but works here; perhaps the same expr but with `method=sqf`? 

The live.sympy.org site is down right now @asmeurer with message: "You have reached a domain that is pending ICANN verification."

File: sympy/sets/fancysets.py [hunk 9]
```diff
@@ -807,6 +832,48 @@ def __getitem__(self, i):
...
+
+        # it must always be an integer for each value in self
+        F = f(self.step*i + self.start)  # for i in range(len(self))
+        if F.is_integer:
+            n = self.size
+            step = 1
+            match = F.match(a*i + b)
+            a = A
+            b = B
+            p = abs(a.p)
```
Comment (smichr) [inline sympy/sets/fancysets.py]: if F = a*j/aq + b/bq for j in range(0, n, 1) then whether the F(j) is always an integer or not we can write F' = j/aq + b/bq for j in range(0, a*n, a). And that's good enough. Any range starting at a point other than 0 will be remapped to one starting at 0. 

- [ ] check what happens to a neg step range

File: sympy/polys/polytools.py (hunk 8 above)
Comment (smichr) [inline sympy/polys/polytools.py]: I added that test. This is ready to go. I'll commit in 24h if there is no objection.

**Pull request merged**